OpenRA.Platforms.Default/ThreadedGraphicsContext.cs

```diff
@@ -0,0 +1,702 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using OpenRA.Graphics;
+
+namespace OpenRA.Platforms.Default
+{
+	/// <summary>
+	/// Creates a dedicated thread for the graphics device. An internal message queue is used to perform actions on the
+	/// device. This allows calls to be enqueued to be processed asynchronously and thus free up the calling thread.
+	/// </summary>
+	sealed class ThreadedGraphicsContext : IGraphicsContext
+	{
+		// PERF: Maintain several object pools to reduce allocations.
+		readonly Stack<Vertex[]> verticesPool = new Stack<Vertex[]>();
+		readonly Stack<Message> messagePool = new Stack<Message>();
+		readonly Queue<Message> messages = new Queue<Message>();
+
+		readonly object syncObject = new object();
+		readonly Thread renderThread;
+		readonly int batchSize;
+		volatile ExceptionDispatchInfo messageException;
+
+		// Delegates that perform actions on the real device.
+		Func<object> doClear;
+		Action doClearDepthBuffer;
+		Action doDisableDepthBuffer;
+		Action doEnableDepthBuffer;
+		Action doDisableScissor;
+		Action doPresent;
+		Func<string> getGLVersion;
+		Func<ITexture> getCreateTexture;
+		Func<object, ITexture> getCreateTextureBitmap;
+		Func<Bitmap> getTakeScreenshot;
+		Func<object, IFrameBuffer> getCreateFrameBuffer;
+		Func<object, IShader> getCreateShader;
+		Func<object, IVertexBuffer<Vertex>> getCreateVertexBuffer;
+		Action<object> doDrawPrimitives;
+		Action<object> doEnableScissor;
+		Action<object> doSetBlendMode;
+
+		public ThreadedGraphicsContext(Sdl2GraphicsContext context, int batchSize)
+		{
+			this.batchSize = batchSize;
+			renderThread = new Thread(RenderThread)
+			{
+				Name = "ThreadedGraphicsDevice RenderThread",
```

Comment:
Should now be `Name = "ThreadedGraphicsContext RenderThread"`. (Could use `typeof` if you like - up to you).